clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java

```diff
@@ -123,4 +125,31 @@ default boolean isSenderConnected(final Instant now) {
         }
     }
 
+    /**
+     * Returns the time until disconnection notification of this downstream message.
+     *
+     * @return A notification if the message contains a TTD value {@link Optional#empty()} otherwise.
+     */
+    default Optional<TimeUntilDisconnectNotification> getTimeUntilDisconnectNotification() {
+
+        final Integer ttd = getTimeTillDisconnect();
+        final Instant creationTime = getCreationTime();
+
+        if (ttd == null) {
+            return Optional.empty();
+        } else if (ttd == 0 || MessageHelper.isDeviceCurrentlyConnected(ttd,
+                creationTime != null ? creationTime.toEpochMilli() : null)) {
+            final String tenantId = getTenantId();
+            final String deviceId = getDeviceId();
+
+            if (tenantId != null && deviceId != null) {
+                final TimeUntilDisconnectNotification notification =
+                        new TimeUntilDisconnectNotification(tenantId, deviceId, ttd,
+                                TimeUntilDisconnectNotification.getReadyUntilInstantFromTtd(ttd, creationTime), creationTime);
```

Comment:
it looks very starnge to pass in ttd and creation time as parameters and then also being required to use the same parameters to compute another one. Shouldn't the constructor take care of that?

Reply:
Absolutely! It's implemented in `TimeUntilDisconnectNotification#fromMessage` which I used as a template, I'll fix it there as well.